daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs

```diff
@@ -672,23 +680,55 @@ runStart sandboxPort (StartNavigator shouldStartNavigator) (OpenBrowser shouldOp
                   then withNavigator
                   else (\_ _ _ f -> f sandboxPh)
 
-runDeploy :: String -> SandboxPort -> IO ()
-runDeploy host sandboxPort = do
+data HostAndPortFlags = HostAndPortFlags { hostM :: Maybe String, portM :: Maybe Int }
+
+withHostAndPortDefaults :: HostAndPortFlags -> IO HostAndPort
+withHostAndPortDefaults HostAndPortFlags{hostM,portM} = do
+    host <- fromMaybeM getProjectLedgerHost hostM
+    port <- fromMaybeM getProjectLedgerPort portM
+    return HostAndPort {host,port}
+
+runListParties :: HostAndPortFlags -> IO ()
+runListParties flags = do
+    hp <- withHostAndPortDefaults flags
+    putStrLn $ "Listing parties at " <> show hp
+    xs <- Ledger.listParties hp
+    when (null xs) $ putStrLn "no parties are known"
+    mapM_ print xs
+    exitSuccess
+
+runAllocateParty :: HostAndPortFlags -> String -> IO ()
+runAllocateParty flags name = do
+    hp <- withHostAndPortDefaults flags
+    putStrLn $ "Checking party allocation at " <> show hp
+    allocatePartyIfRequired hp name
+    exitSuccess
+
+runDeploy :: HostAndPortFlags -> IO ()
+runDeploy flags = do
+    hp <- withHostAndPortDefaults flags
+    putStrLn $ "Deploying to " <> show hp
+    parties <- getProjectParties
+    mapM_ (allocatePartyIfRequired hp) parties
```

Comment:
Is there a way to do a batch allocation?

Reply:
`daml deploy` does batch allocation of everything in the `daml.yaml` config.

Reply:
I meant batch allocation over the Ledger API. This looks like we're checking for each individual party if it already exists and allocate it when it doesn't. That's `O(n)` API calls for allocating `n` parties. I'm wondering if there's a way to do it in `O(1)` API calls? ("No" is a perfectly fine answer.)

Reply:
"No", the API does not support batch allocation.

Reply:
Thanks.